### PR TITLE
Update `Scrollable` on `ScrollBehaviour` change.

### LINKED ISSUE #130793: Scrollable not changing ScrollBehaviour
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Create a `Scrollable` widget and change its `scrollBehavior` property at runtime (using state or hot reload).

### Expected results

Changes to `scrollBehavior` should be applied without completely recreating the `Scrollable`s state.

### Actual results

The observed scroll behavior always stays the same.

### Code sample

This version does not work:
```dart
class _NotWorkingState extends State<NotWorking> {
  bool _enable = true;

  @override
  Widget build(BuildContext context) => MaterialApp(
        home: Scaffold(
          body: Scrollable(
            scrollBehavior: const MaterialScrollBehavior().copyWith(dragDevices: {
              if (_enable) PointerDeviceKind.mouse,
            }),
            viewportBuilder: (context, position) => Viewport(
              offset: position,
              slivers: [
                for (int i = 0; i < 100; i++)
                  SliverToBoxAdapter(
                    child: Text("Item $i"),
                  ),
              ],
            ),
          ),
          floatingActionButton: FloatingActionButton(onPressed: () {
            setState(() {
              _enable = !_enable;
            });
          }),
        ),
      );
}
```

While this one does:
```dart
class _WorkingState extends State<Working> {
  bool _enable = true;

  @override
  Widget build(BuildContext context) => MaterialApp(
        scrollBehavior: const MaterialScrollBehavior().copyWith(dragDevices: {
          if (_enable) PointerDeviceKind.mouse,
        }),
        home: Scaffold(
          body: Scrollable(
            viewportBuilder: (context, position) => Viewport(
              offset: position,
              slivers: [
                for (int i = 0; i < 100; i++)
                  SliverToBoxAdapter(
                    child: Text("Item $i"),
                  ),
              ],
            ),
          ),
          floatingActionButton: FloatingActionButton(onPressed: () {
            setState(() {
              _enable = !_enable;
            });
          }),
        ),
      );
}
```

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[√] Flutter (Channel stable, 3.10.6, on Microsoft Windows [Version 10.0.19045.3208], locale de-DE)
    • Flutter version 3.10.6 on channel stable at C:\Users\...\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f468f3366c (6 days ago), 2023-07-12 15:19:05 -0700
    • Engine revision cdbeda788a
    • Dart version 3.0.6
    • DevTools version 2.23.1

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0) 
    • Android SDK at C:\Users\...\AppData\Local\Android\sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java       
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)     
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Professional 2022 17.6.5) 
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Professional
    • Visual Studio Professional 2022 version 17.6.33829.357
    • Windows 10 SDK version 10.0.22000.0

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.80.1)
    • VS Code at C:\Users\...\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.68.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.3208]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 114.0.5735.199
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 114.0.1823.82

[√] Network resources
    • All expected network resources are available.
```

</details>

### PR DESCRIPTION
**Pull request by chinmoy12c**
Fixes: #130793 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 1]
```diff
@@ -624,6 +624,9 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
   }
 
   bool _shouldUpdatePosition(Scrollable oldWidget) {
+    if (widget.scrollBehavior != null && oldWidget.scrollBehavior != null && widget.scrollBehavior!.shouldNotify(oldWidget.scrollBehavior!)) {
+      return true;
+    }
```
Comment (Dampfwalze) [inline packages/flutter/lib/src/widgets/scrollable.dart:629]: Maybe add condition
```dart
if((widget.scrollBehavior != null) != (oldWidget.scrollBehavior != null)) {
  return true;
}
```

Comment (chinmoy12c) [reply packages/flutter/lib/src/widgets/scrollable.dart:629]: I think in such a case a call to `_updatePosition` is made either way, so it's handled.

**Review by Piinks — DISMISSED** [review]
LGTM! Thank you for fixing this!

**Comment by Piinks** [comment]
Oh whoops! Let me find another reviewer here. :)

File: packages/flutter/test/widgets/scrollable_test.dart [hunk 2]
```diff
@@ -1356,6 +1356,59 @@ void main() {
...
+    expect(getScrollOffset(tester), 0.0);
+    await gesture.moveBy(const Offset(0.0, -200));
+    await tester.pumpAndSettle();
+    expect(getScrollOffset(tester), 0.0);
+
+    // Change state to include mouse pointer device
+    await tester.tap(find.byType(FloatingActionButton));
+    await tester.pump();
+
+    // Gesture should work after state change
```
Comment (justinmc) [inline packages/flutter/test/widgets/scrollable_test.dart:1405]: Nit: Periods at the ends of these comments.

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 1 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/scrollable.dart:629]: I am a bit confused, if `oldWidget.scrollBehavior == null` but `widget.scrollBehavior != null`, which code path will call `_updatePosition`?

**Review by justinmc — DISMISSED** [review]
LGTM 👍 .  Thanks!

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 1 above)
Comment (chinmoy12c) [reply packages/flutter/lib/src/widgets/scrollable.dart:629]: Hey @chunhtai in this case `_updatePosition` is called from `didChangeDependencies`.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/scrollable.dart:629]: @chinmoy12c can you take a look at this? Thanks!

**Comment by Renzo-Olivares** [comment]
Is this ready to be merged?

File: packages/flutter/lib/src/widgets/scrollable.dart (hunk 1 above)
Comment (chinmoy12c) [reply packages/flutter/lib/src/widgets/scrollable.dart:629]: Hey @chunhtai @Piinks sorry for the late reply, I've been away for a while.

@chunhtai you are right, the code does not trigger `didChangeDependencies`, my bad. The actual code that triggers the update is this:
```
ScrollPhysics? newPhysics = widget.physics ?? widget.scrollBehavior?.getScrollPhysics(context);
ScrollPhysics? oldPhysics = oldWidget.physics ?? oldWidget.scrollBehavior?.getScrollPhysics(context);
```

In this case either the `newPhysics` or the `oldPhysics` is null, hence

```
 if (newPhysics?.runtimeType != oldPhysics?.runtimeType) {
        return true;
 }
```

returns true.

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/scrollable.dart:629]: it won't hit this case if both `widget.physics` and `oldWidget.physics` are not null even if 
oldWidget.scrollBehavior == null  and widget.scrollBehavior != null

If it fails to call updatePosition in this case, the `_configuration` will not be update with the newest widget.scrollBehavior.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/scrollable.dart:629]: I've removed the approvals while we sort this out. @chinmoy12c maybe we can add a test case for the case @chunhtai mentioned above here to better figure out this fix? :)

Comment (chinmoy12c) [reply packages/flutter/lib/src/widgets/scrollable.dart:640]: Hey @chunhtai thanks for pointing that out, I've added another test case to check for that behavior.
@Piinks its ready for another review :).

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thank you!

**Pull request merged**